File: 77.组合.go
/*
 * @lc app=leetcode.cn id=77 lang=golang
 *
 * [77] 组合
 */

// @lc code=start
func combine(n int, k int) [][]int {
	ans := [][]int{}
	var backtracking func(res []int) 
	backtracking = func(res []int) {
		if len(res) == k {
			resCopy := make([]int,k)
			copy(resCopy,res)
			ans = append(ans,resCopy)
			return
		}
		if len(res) == 0 {
			for i := 1 ; i <= n ; i++ {
				res = append(res,i)
				backtracking(res)
				res = res[:len(res)-1]
			}
		}else{
			for i := res[len(res)-1]+1 ; i <= n ; i++ {
				res = append(res,i)
				backtracking(res)
				res = res[:len(res)-1]
			}
		}
	}
	backtracking([]int{})
	return ans
}
// @lc code=end

File: 77.组合.py
#
# @lc app=leetcode.cn id=77 lang=python3
#
# [77] 组合
#

# @lc code=start
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        l = []
        nums = [num for num in range(1,n+1)]
        self.backtrack(nums, l, k, res)
        return res


    def backtrack(self, nums: [int], l: [int], k: int, res: [int]):
        if len(l) == k:
            res.append(copy.deepcopy(l))
            return
        
        for i, num in enumerate(nums):
            l.append(num)
            self.backtrack(nums[i+1:], l, k, res)
            l.pop()
# @lc code=end

